File: src/main/java/infinitespire/relics/PuzzleCube.java
package infinitespire.relics;

import com.megacrit.cardcrawl.actions.common.DrawCardAction;
import com.megacrit.cardcrawl.actions.common.RelicAboveCreatureAction;
import com.megacrit.cardcrawl.cards.AbstractCard;
import com.megacrit.cardcrawl.cards.CardGroup;
import com.megacrit.cardcrawl.dungeons.AbstractDungeon;
import com.megacrit.cardcrawl.monsters.AbstractMonster;
import com.megacrit.cardcrawl.vfx.cardManip.ShowCardBrieflyEffect;
import infinitespire.InfiniteSpire;
import infinitespire.abstracts.Relic;
import infinitespire.patches.AbstractCardPatch;

import java.util.ArrayList;

public class PuzzleCube extends Relic {

	public static final String ID = InfiniteSpire.createID("PuzzleCube");
	private AbstractCard card;

	public PuzzleCube(){
		super(ID, "puzzlecube", RelicTier.COMMON, LandingSound.SOLID);
	}

	@Override
	public void atTurnStart() {
		CardGroup drawPile = AbstractDungeon.player.drawPile;
		CardGroup allCards = new CardGroup(CardGroup.CardGroupType.UNSPECIFIED);

		allCards.group.addAll(AbstractDungeon.player.discardPile.group);
		allCards.group.addAll(AbstractDungeon.player.drawPile.group);
		allCards.group.addAll(AbstractDungeon.player.hand.group);
		allCards.group.addAll(AbstractDungeon.player.exhaustPile.group);

		for(AbstractCard card : allCards.group){
			AbstractCardPatch.Field.isPuzzleCubeCard.set(card, false);
		}

		if(drawPile.size() > 0){
			this.flash();
			AbstractCard card = getCard(drawPile);
			this.card = card;
			AbstractCard cardCopy = card.makeStatEquivalentCopy();
			AbstractCardPatch.Field.isPuzzleCubeCard.set(cardCopy, true);
			AbstractDungeon.effectsQueue.add(new ShowCardBrieflyEffect(cardCopy));
			AbstractCardPatch.Field.isPuzzleCubeCard.set(card, true);
		}
	}

	private AbstractCard getCard(CardGroup group){
		ArrayList<AbstractCard> cards = new ArrayList<>(group.group);
		cards.removeIf(
			abstractCard ->
				abstractCard.type == AbstractCard.CardType.CURSE ||
				abstractCard.type == AbstractCard.CardType.STATUS);

		return cards.get(AbstractDungeon.cardRandomRng.random(cards.size() - 1));
	}

	@Override
	public void onPlayCard(AbstractCard c, AbstractMonster m) {
		if(AbstractCardPatch.Field.isPuzzleCubeCard.get(c)){
			AbstractDungeon.actionManager.addToBottom(new RelicAboveCreatureAction(AbstractDungeon.player, this));
			AbstractDungeon.actionManager.addToBottom(new DrawCardAction(AbstractDungeon.player, 1));
			AbstractCardPatch.Field.isPuzzleCubeCard.set(card, false);
		}
	}
}

File: src/main/kotlin/com/blanktheevil/infinitespire/relics/PuzzleCube.kt
package com.blanktheevil.infinitespire.relics

import com.badlogic.gdx.graphics.Color
import com.blanktheevil.infinitespire.cards.utils.inPuzzleCube
import com.blanktheevil.infinitespire.extensions.actionManager
import com.blanktheevil.infinitespire.extensions.getRandomItem
import com.blanktheevil.infinitespire.extensions.makeID
import com.blanktheevil.infinitespire.extensions.player
import com.megacrit.cardcrawl.actions.common.DrawCardAction
import com.megacrit.cardcrawl.actions.common.RelicAboveCreatureAction
import com.megacrit.cardcrawl.cards.AbstractCard
import com.megacrit.cardcrawl.cards.CardGroup
import com.megacrit.cardcrawl.dungeons.AbstractDungeon
import com.megacrit.cardcrawl.monsters.AbstractMonster
import com.megacrit.cardcrawl.vfx.cardManip.ShowCardBrieflyEffect
import kotlin.streams.toList

class PuzzleCube : Relic(ID, IMG, TIER, SOUND) {
  companion object {
    val ID = "PuzzleCube".makeID()
    private const val IMG = "puzzlecube"
    private val TIER = RelicTier.COMMON
    private val SOUND = LandingSound.SOLID
    private val DEFAULT_BLUE = Color(0.2f, 0.9f, 1f, 0.25f)
  }

  private var selectedCard: AbstractCard? = null

  override fun atTurnStart() {
    val allCards = CardGroup(CardGroup.CardGroupType.UNSPECIFIED).also {
      with(it.group) {
        addAll(player.discardPile.group)
        addAll(player.drawPile.group)
        addAll(player.hand.group)
        addAll(player.exhaustPile.group)
      }
    }

    allCards.group.forEach {
      it.inPuzzleCube = false
      it.glowColor = DEFAULT_BLUE.cpy()
    }

    if (player.drawPile.size() > 0) {
      flash()
      getCard(player.drawPile).apply {
        if (this != null) {
          selectedCard = this
          val copy = this.makeStatEquivalentCopy()
          copy.inPuzzleCube = true
          this.inPuzzleCube = true
          copy.glowColor = Color.GOLD.cpy()
          this.glowColor = Color.GOLD.cpy()
          AbstractDungeon.effectsQueue.add(ShowCardBrieflyEffect(copy))
        }
      }
    }
  }

  override fun onPlayCard(c: AbstractCard, m: AbstractMonster?) {
    if (c.inPuzzleCube) {
      actionManager.addToBottom(RelicAboveCreatureAction(player, this))
      actionManager.addToBottom(DrawCardAction(player, 1))
      c.inPuzzleCube = false
    }
  }

  private fun getCard(group: CardGroup): AbstractCard? {
    return group.group.stream()
      .filter { it.type != AbstractCard.CardType.CURSE && it.type != AbstractCard.CardType.STATUS }
      .toList()
      .getRandomItem(AbstractDungeon.cardRandomRng)
  }
}